core/src/bytecode/typecheck/mod.rs

```diff
@@ -1765,46 +1793,60 @@ impl<'ast> Walk<'ast> for FieldDef<'ast> {
     }
 }
 
+impl<'ast> Walk<'ast> for &'ast LetBinding<'ast> {
+    fn walk<V: TypecheckVisitor<'ast>>(
+        self,
+        state: &mut State<'ast, '_>,
+        ctxt: Context<'ast>,
+        visitor: &mut V,
+    ) -> Result<(), TypecheckError> {
+        walk_with_annot(
+            state,
+            ctxt,
+            visitor,
+            &self.metadata.annotation,
+            Some(&self.value),
+        )
+    }
+}
+
 /// Walk an annotated term, either via [crate::term::record::FieldMetadata], or via a standalone
 /// type and contracts annotation. A type annotation switches the typechecking mode to _enforce_.
-fn walk_with_annot<'ast, V: TypecheckVisitor<'ast>>(
+fn walk_with_annot<'ast, 'a, S: AnnotSeqRef<'ast>, V: TypecheckVisitor<'ast>>(
     state: &mut State<'ast, '_>,
     mut ctxt: Context<'ast>,
     visitor: &mut V,
-    annot: &Annotation<'ast>,
-    value: Option<&Ast<'ast>>,
+    annots: S,
+    value: Option<&'ast Ast<'ast>>,
 ) -> Result<(), TypecheckError> {
-    annot
+    annots
         .iter()
         .try_for_each(|ty| ty.walk(state, ctxt.clone(), visitor))?;
 
-    match (annot, value) {
-        (Annotation { typ: Some(ty2), .. }, Some(value)) => {
+    let typ = annots.typ();
+
+    match (typ, value) {
+        (Some(ty2), Some(value)) => {
             let uty2 = UnifType::from_type(ty2.clone(), &ctxt.term_env);
             value.check(state, ctxt, visitor, uty2)
         }
-        (
-            Annotation {
-                typ: None,
-                contracts,
-            },
-            Some(value),
-        ) => {
+        (None, Some(value)) => {
             // If we see a function annotated with a function contract, we can get the type of the
             // arguments for free. We use this information both for typechecking (you could see it
             // as an extension of the philosophy of apparent types, but for function arguments
             // instead of let-bindings) and for the LSP, to provide better type information and
             // completion.
             if let Node::Fun { args, body } = value.node {
                 // We look for the first contract of the list that is a function contract.
-                let domains = contracts.iter().find_map(|c| {
-                    if let TypeF::Arrow(mut domain, _) = &c.typ {
+                let domains = annots.contracts().find_map(|c| {
+                    if let TypeF::Arrow(domain, mut codomain) = &c.typ {
                         let mut domains =
                             vec![UnifType::from_type((*domain).clone(), &ctxt.term_env)];
 
-                        while let TypeF::Arrow(next_domain, _) = &domain.typ {
-                            domains.push(UnifType::from_type((*domain).clone(), &ctxt.term_env));
-                            domain = next_domain;
+                        while let TypeF::Arrow(next_domain, next_codomain) = &codomain.typ {
```

Comment:
I don't follow the details here, but this looks like an actual change and not a refactor. Is this fixing some bug, and if so should there be a test for it?

Reply:
Yes, the previous code was wrong: if you have a multi-parameter function with a contract `Foo -> Bar -> Baz`, to get type hints for the parameter, you need to look into the domain of the  contract but then iterate on the _codomain_. But the code iterated on the domain. I believe this wasn't used in statically typed code enough to create a typechecking bug, but it's visible on multi-parameter function with contracts in the LSP, where the hovering is off. I'll try to reconstruct a simple example and add it to the tests, yeah.

Reply:
In fact I'm dumb, I  can't add a test in this PR or it will be failing (since we still use the old typechecker). It'll need to land in a subsequent PR.